File: frontend/lib/models/endereco_model.dart
class Endereco {
  final int id;
  final String cep;
  final String uf;
  final String cidade;
  final String logradouro;
  final String? complemento;
  final String bairro;
  final int? numero;

  Endereco({
    required this.id,
    required this.cep,
    required this.uf,
    required this.cidade,
    required this.logradouro,
    this.complemento,
    required this.bairro,
    required this.numero,
  });

  factory Endereco.fromJson(Map<String, dynamic> json) {
    return Endereco(
      id: json['ID'] ?? 0, // Use a default value of 0 if ID is null
      cep: json['CEP'] ?? '', // Handle missing field with a default value
      uf: json['UF'] ?? '', // Handle missing field with a default value
      cidade: json['CIDADE'] ?? '', // Handle missing field with a default value
      logradouro:
          json['LOGRADOURO'] ?? '', // Handle missing field with a default value
      complemento: json['COMPLEMENTO'], // Nullable
      bairro: json['BAIRRO'] ?? '', // Handle missing field with a default value
      numero: json['NUMERO'], // Handle missing field with a default value
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'ID': id,
      'CEP': cep,
      'UF': uf,
      'CIDADE': cidade,
      'LOGRADOURO': logradouro,
      'COMPLEMENTO': complemento,
      'BAIRRO': bairro,
      'NUMERO': numero,
    };
  }
}

File: backend/models/enderecoModel.js
const { DataTypes } = require('sequelize');
const db = require('../database/db');

const Endereco = db.define('Endereco', {
  ID: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  CEP: {
    type: DataTypes.STRING(8), // Validated as 8-digit string
    allowNull: false,
    validate: {
      is: /^\d{8}$/  // Ensures the CEP is exactly 8 digits
    }
  },
  UF: {
    type: DataTypes.STRING(2), // State abbreviation (2 letters)
    allowNull: false
  },
  CIDADE: {
    type: DataTypes.STRING(255),
    allowNull: false
  },
  LOGRADOURO: {
    type: DataTypes.STRING(255),
    allowNull: false
  },
  COMPLEMENTO: {
    type: DataTypes.STRING(255),
    allowNull: true
  },
  BAIRRO: {
    type: DataTypes.STRING(255),
    allowNull: true
  },
  NUMERO: {
    type: DataTypes.INTEGER,
    allowNull: true
  }
}, {
  tableName: 'ENDERECOS',
  timestamps: true, // Automatically adds `CRIADO_EM` and `ATUALIZADO_EM` fields
  createdAt: 'CRIADO_EM',
  updatedAt: 'ATUALIZADO_EM'
});

module.exports = Endereco;